File: src/bin/cli/cmd/set.rs
use crate::cli::OutputFormat;
use crate::dto::SetResultDto;
use crate::json_output::{output_error, output_success};
use crate::utils;
use lib::backup::ports::BackupService;
use lib::backup::service::SubRipBackupService;
use lib::subtitle::model::{Subtitle, SubtitleText, SubtitleTimestamp, SubtitleUpdate};
use lib::subtitle::ports::SubtitleService;
use lib::subtitle::service::SubRipService;
use log::{debug, error, info};

pub fn handle(
    file: &str,
    index: u32,
    start: Option<String>,
    end: Option<String>,
    text: Option<String>,
    dry_run: bool,
    format: &OutputFormat,
) -> anyhow::Result<()> {
    info!("setting subtitle {} in file: {}", index, file);

    if start.is_none() && end.is_none() && text.is_none() {
        output_error(
            format,
            "no_fields_to_update",
            "At least one of --start, --end, or --text must be specified",
            None,
        );
        return Err(anyhow::anyhow!(""));
    }

    // Validate text input
    if let Some(ref t) = text {
        utils::reject_control_chars(t, "text")?;
    }

    let resolved = utils::resolve_existing_path(file)?;

    let start_timestamp = if let Some(start_str) = start {
        debug!("parsing start timestamp: {}", start_str);
        let duration = Subtitle::parse_timestamp(&start_str)
            .map_err(|e| anyhow::anyhow!("invalid start timestamp: {}", e))?;
        Some(
            SubtitleTimestamp::try_new(duration)
                .map_err(|e| anyhow::anyhow!("invalid start timestamp value: {}", e))?,
        )
    } else {
        None
    };

    let end_timestamp = if let Some(end_str) = end {
        debug!("parsing end timestamp: {}", end_str);
        let duration = Subtitle::parse_timestamp(&end_str)
            .map_err(|e| anyhow::anyhow!("invalid end timestamp: {}", e))?;
        Some(
            SubtitleTimestamp::try_new(duration)
                .map_err(|e| anyhow::anyhow!("invalid end timestamp value: {}", e))?,
        )
    } else {
        None
    };

    let subtitle_text = if let Some(text_str) = text {
        debug!("validating text (length: {})", text_str.len());
        Some(SubtitleText::try_new(text_str).map_err(|e| anyhow::anyhow!("invalid text: {}", e))?)
    } else {
        None
    };

    let update = SubtitleUpdate {
        start_time: start_timestamp,
        end_time: end_timestamp,
        text: subtitle_text,
    };

    if dry_run {
        info!("dry-run mode, previewing changes");
        let service = SubRipService::new(resolved.base_dir);

        // Check subtitle exists and show before/after
        match service.get_by_id(&resolved.filename, index) {
            Ok(Some(existing)) => {
                let updated = update
                    .apply_to(&existing)
                    .map_err(|e| anyhow::anyhow!("invalid update: {}", e))?;

                let mut fields = Vec::new();
                if update.start_time.is_some() {
                    fields.push("start_time".to_string());
                }
                if update.end_time.is_some() {
                    fields.push("end_time".to_string());
                }
                if update.text.is_some() {
                    fields.push("text".to_string());
                }

                let dto = SetResultDto {
                    index,
                    fields_updated: fields,
                    backup_path: "N/A (dry-run)".into(),
                    dry_run: true,
                };

                output_success(format, &dto, || {
                    println!("Dry-run: Subtitle {} would be updated", index);
                    println!();
                    println!("Before:");
                    println!("{}", existing);
                    println!();
                    println!("After:");
                    println!("{}", updated);
                });
                return Ok(());
            }
            Ok(None) => {
                output_error(
                    format,
                    "subtitle_not_found",
                    &format!("Subtitle with index {} not found in file", index),
                    None,
                );
                return Err(anyhow::anyhow!(""));
            }
            Err(e) => {
                let cli_err = utils::format_subtitle_error(&e, file);
                output_error(
                    format,
                    &cli_err.code,
                    &cli_err.message,
                    cli_err.hint.as_deref(),
                );
                return Err(anyhow::anyhow!(""));
            }
        }
    }

    let backup_service = SubRipBackupService::new();
    let backup_result = backup_service.create_backup(&resolved.full_path);

    let backup_path = match backup_result {
        Ok(Some(path)) => {
            debug!("backup created: {}", path);
            path
        }
        Ok(None) => {
            error!("file does not exist, cannot update");
            output_error(
                format,
                "file_not_found",
                &format!("File does not exist: {}", file),
                None,
            );
            return Err(anyhow::anyhow!(""));
        }
        Err(e) => {
            error!("failed to create backup: {}", e);
            output_error(
                format,
                "backup_failed",
                &format!("Failed to create backup: {}", e),
                None,
            );
            return Err(anyhow::anyhow!(""));
        }
    };

    let service = SubRipService::new(resolved.base_dir);

    debug!("updating subtitle {}...", index);
    match service.set(&resolved.filename, index, update) {
        Ok(report) => {
            info!("subtitle {} updated successfully", index);

            let dto = SetResultDto {
                index,
                fields_updated: report.fields_updated.clone(),
                backup_path: backup_path.clone(),
                dry_run: false,
            };

            output_success(format, &dto, || {
                println!("✓ Subtitle {} updated successfully", index);
                println!();
                println!("Backup: {}", backup_path);
                println!("Fields updated: {}", report.fields_updated.join(", "));
            });

            Ok(())
        }
        Err(e) => {
            let cli_err = utils::format_subtitle_error(&e, file);
            output_error(
                format,
                &cli_err.code,
                &cli_err.message,
                cli_err.hint.as_deref(),
            );
            Err(anyhow::anyhow!(""))
        }
    }
}

File: internal/cli/cmd/set.go
// Package cmd holds the command handlers for the submarine CLI. Each handler is
// a 1-to-1 port of the corresponding Rust `src/bin/cli/cmd/*.rs` module.
package cmd

import (
	"errors"
	"fmt"
	"log/slog"
	"strings"

	"github.com/lebe-dev/submarine/internal/cli"
	"github.com/lebe-dev/submarine/pkg/backup"
	"github.com/lebe-dev/submarine/pkg/subtitle"
)

// HandleSet is a 1-to-1 port of the Rust `cmd::set::handle`.
func HandleSet(
	file string,
	index uint32,
	start *string,
	end *string,
	text *string,
	dryRun bool,
	format cli.OutputFormat,
) error {
	slog.Info(fmt.Sprintf("setting subtitle %d in file: %s", index, file))

	if start == nil && end == nil && text == nil {
		cli.OutputError(
			format,
			"no_fields_to_update",
			"At least one of --start, --end, or --text must be specified",
			nil,
		)
		return errors.New("")
	}

	// Validate text input
	if text != nil {
		if err := cli.RejectControlChars(*text, "text"); err != nil {
			return err
		}
	}

	resolved, err := cli.ResolveExistingPath(file)
	if err != nil {
		return err
	}

	var startTimestamp *subtitle.SubtitleTimestamp
	if start != nil {
		slog.Debug(fmt.Sprintf("parsing start timestamp: %s", *start))
		duration, err := subtitle.ParseTimestamp(*start)
		if err != nil {
			return fmt.Errorf("invalid start timestamp: %s", err)
		}
		ts, err := subtitle.NewSubtitleTimestamp(duration)
		if err != nil {
			return fmt.Errorf("invalid start timestamp value: %s", err)
		}
		startTimestamp = &ts
	}

	var endTimestamp *subtitle.SubtitleTimestamp
	if end != nil {
		slog.Debug(fmt.Sprintf("parsing end timestamp: %s", *end))
		duration, err := subtitle.ParseTimestamp(*end)
		if err != nil {
			return fmt.Errorf("invalid end timestamp: %s", err)
		}
		ts, err := subtitle.NewSubtitleTimestamp(duration)
		if err != nil {
			return fmt.Errorf("invalid end timestamp value: %s", err)
		}
		endTimestamp = &ts
	}

	var subtitleText *subtitle.SubtitleText
	if text != nil {
		slog.Debug(fmt.Sprintf("validating text (length: %d)", len(*text)))
		st, err := subtitle.NewSubtitleText(*text)
		if err != nil {
			return fmt.Errorf("invalid text: %s", err)
		}
		subtitleText = &st
	}

	update := subtitle.SubtitleUpdate{
		StartTime: startTimestamp,
		EndTime:   endTimestamp,
		Text:      subtitleText,
	}

	if dryRun {
		slog.Info("dry-run mode, previewing changes")
		service := subtitle.NewSubRipService(resolved.BaseDir)

		// Check subtitle exists and show before/after
		existing, err := service.GetByID(resolved.Filename, index)
		if err != nil {
			var subErr *subtitle.SubtitleError
			if errors.As(err, &subErr) {
				cliErr := cli.FormatSubtitleError(subErr, file)
				cli.OutputError(
					format,
					cliErr.Code,
					cliErr.Message,
					cliErr.Hint,
				)
				return errors.New("")
			}
			return err
		}

		if existing == nil {
			cli.OutputError(
				format,
				"subtitle_not_found",
				fmt.Sprintf("Subtitle with index %d not found in file", index),
				nil,
			)
			return errors.New("")
		}

		updated, err := update.ApplyTo(*existing)
		if err != nil {
			return fmt.Errorf("invalid update: %s", err)
		}

		var fields []string
		if update.StartTime != nil {
			fields = append(fields, "start_time")
		}
		if update.EndTime != nil {
			fields = append(fields, "end_time")
		}
		if update.Text != nil {
			fields = append(fields, "text")
		}

		dto := cli.SetResultDto{
			Index:         index,
			FieldsUpdated: fields,
			BackupPath:    "N/A (dry-run)",
			DryRun:        true,
		}

		cli.OutputSuccess(format, dto, func() {
			fmt.Printf("Dry-run: Subtitle %d would be updated\n", index)
			fmt.Println()
			fmt.Println("Before:")
			fmt.Printf("%s\n", existing.String())
			fmt.Println()
			fmt.Println("After:")
			fmt.Printf("%s\n", updated.String())
		})
		return nil
	}

	backupService := backup.NewSubRipBackupService()
	backupResult, backupErr := backupService.CreateBackup(resolved.FullPath)

	var backupPath string
	switch {
	case backupErr != nil:
		slog.Error(fmt.Sprintf("failed to create backup: %s", backupErr))
		cli.OutputError(
			format,
			"backup_failed",
			fmt.Sprintf("Failed to create backup: %s", backupErr),
			nil,
		)
		return errors.New("")
	case backupResult == nil:
		slog.Error("file does not exist, cannot update")
		cli.OutputError(
			format,
			"file_not_found",
			fmt.Sprintf("File does not exist: %s", file),
			nil,
		)
		return errors.New("")
	default:
		slog.Debug(fmt.Sprintf("backup created: %s", *backupResult))
		backupPath = *backupResult
	}

	service := subtitle.NewSubRipService(resolved.BaseDir)

	slog.Debug(fmt.Sprintf("updating subtitle %d...", index))
	report, err := service.Set(resolved.Filename, index, update)
	if err != nil {
		var subErr *subtitle.SubtitleError
		if errors.As(err, &subErr) {
			cliErr := cli.FormatSubtitleError(subErr, file)
			cli.OutputError(
				format,
				cliErr.Code,
				cliErr.Message,
				cliErr.Hint,
			)
			return errors.New("")
		}
		return err
	}

	slog.Info(fmt.Sprintf("subtitle %d updated successfully", index))

	dto := cli.SetResultDto{
		Index:         index,
		FieldsUpdated: report.FieldsUpdated,
		BackupPath:    backupPath,
		DryRun:        false,
	}

	cli.OutputSuccess(format, dto, func() {
		fmt.Printf("✓ Subtitle %d updated successfully\n", index)
		fmt.Println()
		fmt.Printf("Backup: %s\n", backupPath)
		fmt.Printf("Fields updated: %s\n", strings.Join(report.FieldsUpdated, ", "))
	})

	return nil
}
